src/Jellyfin.Networking/HappyEyeballs/HttpClientExtension.cs

```diff
@@ -0,0 +1,119 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using System.IO;
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Jellyfin.Networking.HappyEyeballs;
+
+/// <summary>
+/// Defines the <see cref="HttpClientExtension"/> class.
+///
+/// Implementation taken from https://github.com/ppy/osu-framework/pull/4191 .
+/// </summary>
+public static class HttpClientExtension
+{
+    /// <summary>
+    /// Gets or sets a value indicating whether the client should use IPv6.
+    /// </summary>
+    public static bool UseIPv6 { get; set; } = true;
+
+    /// <summary>
+    /// Implements the httpclient callback method.
+    /// </summary>
+    /// <param name="context">The <see cref="SocketsHttpConnectionContext"/> instance.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> instance.</param>
+    /// <returns>The http steam.</returns>
+    public static async ValueTask<Stream> OnConnect(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
+    {
+        if (!UseIPv6)
+        {
+            return await AttemptConnection(AddressFamily.InterNetwork, context, cancellationToken).ConfigureAwait(false);
+        }
+
+        using var cancelIPv6 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        var tryConnectAsyncIPv6 = AttemptConnection(AddressFamily.InterNetworkV6, context, cancelIPv6.Token);
+
+        // GetAwaiter().GetResult() is used instead of .Result as this results in improved exception handling.
+        // The tasks have already been completed.
+        // See https://github.com/dotnet/corefx/pull/29792/files#r189415885 for more details.
+        if (await Task.WhenAny(tryConnectAsyncIPv6, Task.Delay(200, cancelIPv6.Token)).ConfigureAwait(false) == tryConnectAsyncIPv6 && tryConnectAsyncIPv6.IsCompletedSuccessfully)
+        {
+            await cancelIPv6.CancelAsync().ConfigureAwait(false);
+            return tryConnectAsyncIPv6.GetAwaiter().GetResult();
```

Comment:
As its a `Task` not a `ValueTask` its fine to `await` them anyway even if already completed. So either make `AttemptConnection` return a `ValueTask` (which would be the correct usage for it) and/or await it here anyway.

Reply:
All of this code is being moved w/ as few changes as possible to make it compile, so any further changes aren't relevant for this PR.

Reply:
await should always be considered for 3 reasons:
- Change of implementation. When the underlying implementation changes this could be a real task
- Again, that why a ValueTask should be used here. ValueTask represent exactly this behavior.
- Syntactic correctness. If we are in an async method, we should always await tasks.

Reply:
I prefer to do what Stephen Toub says: https://github.com/dotnet/corefx/pull/29792/files#r189415885. He seems like a smart enough guy :)

Reply:
Understandable, i read the whole discussion before starting this :-D

But i would argue that even readability in our, way less critical scenario is more important then 5 bytes and 1ms additional time. 
Tbo i dont know why he did not even considered using ValueTask

Reply:
Can't use a ValueTask in Task.WhenAny though, so how would you wait for a result but only for 200ms?

Reply:
As its just one task you want to wait for 200ms, i would just use `await tryConnectAsyncIPv6.WaitAsync(200).ConfigureAwait`

Reply:
There is no WaitAsync for ValueTask :)

Reply:
Oh really? Then i have to apologise :-)